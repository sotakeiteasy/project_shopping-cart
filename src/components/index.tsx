import Header from "./Header/Header";
import ProductCard from "./ProductCard/ProductCard";
export { Header, ProductCard };
