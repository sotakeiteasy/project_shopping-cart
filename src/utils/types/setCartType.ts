import { Product } from "./product";
export type SetCartType = React.Dispatch<React.SetStateAction<Product[]>>;
