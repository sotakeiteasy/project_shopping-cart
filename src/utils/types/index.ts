import { Product } from "./product";
import { SetCartType } from "./setCartType";

export type { Product, SetCartType };
