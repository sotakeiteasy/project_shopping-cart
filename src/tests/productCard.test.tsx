import { render, screen } from '@testing-library/react'
import { ProductCard } from '@/components'
import { Product } from '@/utils/types'
import userEvent from '@testing-library/user-event'

describe("renders ProductCart", () => {
    it('snapshot productCard', () => {
        const Product: Product = {
            id: 1,
            title: 'test',
            price: 20,
            description: "test",
            category: "test",
            image: "test",
            rating: { count: 2, rate: 5 },
        }

        const {container} = render(<ProductCard product={Product} countCart={2} />)
        expect(container).toMatchSnapshot()       
    })

    it('count does not shows when count === 0', () => {
        const Product: Product = {
            id: 1,
            title: 'test',
            price: 20,
            description: "test",
            category: "test",
            image: "test",
            rating: { count: 2, rate: 5 },
        }

        render(<ProductCard product={Product} countCart={0} />)
        expect(screen.queryByRole('button', {name: '+'})).not.toBeInTheDocument()     
    })

    it('count shows when count > 0', () => {
        const Product: Product = {
            id: 1,
            title: 'test',
            price: 20,
            description: "test",
            category: "test",
            image: "test",
            rating: { count: 2, rate: 5 },
        }

        render(<ProductCard product={Product} countCart={1} />)
        expect(screen.getByRole('button', {name: '+'})).toBeInTheDocument()     
    })

    it('increase and decrease count', async () => {
        const user = userEvent.setup()
        const Product: Product = {
            id: 1,
            title: 'test',
            price: 20,
            description: "test",
            category: "test",
            image: "test",
            rating: { count: 2, rate: 5 },
        }

        render(<ProductCard product={Product} countCart={0} />)
        expect(screen.queryByText('1')).not.toBeInTheDocument();

        const firstAddBtn = screen.getByRole('button', { name: /add to cart/i })
        await user.click(firstAddBtn)
        expect(screen.getByText('1')).toBeInTheDocument();

        const addBtn = screen.getByRole('button', { name: '+' })
        await user.click(addBtn)
        expect(screen.getByText('2')).toBeInTheDocument();
        await user.click(addBtn)
        expect(screen.getByText('3')).toBeInTheDocument();

        const removeBtn = screen.getByRole('button', { name: '-' })
        await user.click(removeBtn)
        expect(screen.getByText('2')).toBeInTheDocument();
        await user.click(removeBtn)
        expect(screen.getByText('1')).toBeInTheDocument();
    })

    it("shows decrease and increase buttons only if count > 0", () => {
        const Product: Product = {
            id: 1,
            title: 'test',
            price: 20,
            description: "test",
            category: "test",
            image: "test",
            rating: { count: 2, rate: 5 },
        }
        render(<ProductCard product={Product} countCart={0} />)
        expect(screen.queryByRole('button', { name: '+' })).not.toBeInTheDocument()
        expect(screen.queryByRole('button', { name: '-' })).not.toBeInTheDocument()
    })

    it("change text when count > 0", async () => {
        const user = userEvent.setup()
        const Product: Product = {
            id: 1,
            title: 'test',
            price: 20,
            description: "test",
            category: "test",
            image: "test",
            rating: { count: 2, rate: 5 },
        }
        render(<ProductCard product={Product} countCart={0} />)
        const addToCartBtn = screen.getByText(/add to cart/i)
        
        await user.click(addToCartBtn)

        expect(addToCartBtn.textContent).toMatch(/in cart/i)
    })
})

