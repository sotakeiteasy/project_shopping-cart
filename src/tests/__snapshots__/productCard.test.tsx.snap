// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders ProductCart > snapshot productCard 1`] = `
<div>
  <div
    class="_productCard_fbb629"
  >
    <div
      class="_imageContainer_fbb629"
    >
      <img
        alt="image of test"
        src="test"
      />
    </div>
    <h3>
      test
    </h3>
    <p>
      5
      ‚≠ê/ 
      2
    </p>
    <span
      class="_price_fbb629"
    >
      20
      $
    </span>
    <span
      class="_buttons_fbb629"
    >
      <button
        class="_buyButton_fbb629"
      >
        In cart
      </button>
      <button
        class="_numberBtn_fbb629"
      >
        -
      </button>
      2
      <button
        class="_numberBtn_fbb629"
      >
        +
      </button>
    </span>
  </div>
</div>
`;
